### ISSUE TITLE
Scaled centred sprites are skewed with pixel snap

### ISSUE THREAD

**Issue by jusw85** [godotengine/godot#60443]
### Godot version

3.4.4

### System information

Arch Linux

### Issue description

Related #42985

quality/2d/use_pixel_snap=true
2d/snapping/use_gpu_pixel_snap=true

Import a 1x1 sprite
centred=true
scale=(1, 1)

![capture_005_23042022_024031](https://user-images.githubusercontent.com/3406652/164775001-b1bb9edc-2a24-4007-9e71-151ad136ebf9.jpg)

As expected.

centred=true
scale=(32,32)

![capture_004_23042022_024014](https://user-images.githubusercontent.com/3406652/164775172-e9b3df68-742f-457f-8ccd-9d7f87f3da55.jpg)

Is this intended behaviour? The sprite origin is on the bottom right of the sprite instead of the centre of the scaled sprite. I expected the 32x32 sprite to be centred, with sprite origin at it's centre.

This issue does not happen with textures with even dimensions.

### Steps to reproduce

-

### Minimal reproduction project

[TestPixelSnap2.zip](https://github.com/godotengine/godot/files/8543251/TestPixelSnap2.zip)

**Comment by Calinou** [comment]
cc @lawnjelly

**Comment by lawnjelly** [comment]
Have limited time today, but this seems to have been the behaviour back to at least 3.2, possibly since GPU snap was first added.

I was kind of guessing that possibly the snap was being done before the scale was applied, but that doesn't seem to be the case in the shader (I may be missing something though). Anyone else that wants to look into this is welcome BTW.

**Comment by jusw85** [comment]
> Maybe this issue should be closed as a duplicate?

Yes, I see this issue as more of the root cause, while #42985 is a bug caused by this. Perhaps #42985 could be closed while this issue is used as the tracking issue.

**Comment by kleonc** [comment]
I think this shouldn't be there at all:
https://github.com/godotengine/godot/blob/3ba980379dfb72f5d238640835d8e77c76cf3c99/scene/2d/sprite.cpp#L102-L104
GPU pixel snapping should be applied after all transformations are applied (so in the shader), there's no point to apply it locally (`r_dst_rect` in `Sprite::_get_rects` is local). Transforms are applied later when drawing (by the `VisualServer` based on the `CanvasItem` passed to `Texture::draw_rect_region`):
https://github.com/godotengine/godot/blob/3ba980379dfb72f5d238640835d8e77c76cf3c99/scene/2d/sprite.cpp#L118-L133
